### PR TITLE
Feature: Modify User Status Via API

### PR DESCRIPTION
**Pull request by szymonbalasz**
## Product Description
This change allows activating and deactivating mobile workers via the CommCare HQ API. It enables toggling the active status of mobile worker accounts through API endpoints.

## Technical Summary
The changes allow activating and deactivating mobile workers through the CommCare HQ API by adding new endpoints to the UserResource API resource:

1. POST /api/v0.5/user/{user_id}/activate/: Activates the specified mobile worker
2. POST /api/v0.5/user/{user_id}/deactivate/: Deactivates the specified mobile worker

The implementation leverages existing code patterns and extracts shared logic into reusable methods. Permissions checks ensure only users with appropriate privileges can perform these actions.
Two new tests were added in `corehq/apps/api/tests/user_resources.py` to verify the activate and deactivate functionality, following the existing test patterns.

## Feature Flag
Not specific to any feature flag

### Safety story
The changes were tested locally to verify the activate/deactivate functionality works as expected through the API. Existing permission checks were utilized to restrict access. The implementation follows established coding patterns in the codebase to limit risk.

Since this modifies the status of existing mobile worker data, caution should be exercised when activating/deactivating users. As far as I can tell, existing permission checks are in play.

### Automated test coverage

The added tests in `corehq/apps/api/tests/user_resources.py` verify that:

1. Mobile workers can be successfully activated through the API
2. Mobile workers can be successfully deactivated through the API
3. Deactivation is blocked for location-based mobile workers
4. `UserHistory` entities are created for the modification action

These tests would catch regressions related to the activate/deactivate API functionality.

### QA Plan

1. Verify that a mobile worker can be activated through the API by an authorized user
2. Verify that a mobile worker can be deactivated through the API by an authorized user
3. Verify that deactivation is blocked for location-based mobile workers
4. Verify that unauthorized users cannot access the activate/deactivate API endpoints

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Please let me know if you would like me to modify or expand the PR description in any way. I tried to capture the key aspects based on the information provided, but feel free to adapt it as needed.

### CODE REVIEW

File: corehq/apps/api/resources/v0_5.py [hunk 1]
```diff
@@ -326,6 +331,40 @@ def _update(cls, bundle, user_change_logger=None):
...
+    @location_safe
+    def deactivate_user(self, request, **kwargs):
+        return self._modify_user_status(request, **kwargs, active=False)
+
+    def _modify_user_status(self, request, active, **kwargs):
+        self.is_authenticated(request)
+
+        try:
+            user = CommCareUser.get(kwargs['pk'])
+            assert user.domain == kwargs['domain']
```
Comment (snopoke) [inline corehq/apps/api/resources/v0_5.py:355]: This should rather be done using existing methods (and outside the 'try/catch'):
```
user = CommCareUser.get_by_user_id(kwargs['pk'], kwargs["domain"])
if not user:
    raise NotFound()
```

File: corehq/apps/api/tests/user_resources.py [hunk 2]
```diff
@@ -341,6 +341,50 @@ def test_update_fails(self):
...
+        response = self._assert_auth_post_resource(activate_url, json.dumps({}), content_type='application/json',
+                                                   method='POST')
+        updated_user = CommCareUser.get(user.get_id)
+        changes = UserHistory.objects.filter(user_id=user.get_id, changed_via=USER_CHANGE_VIA_API)
+
+        self.assertEqual(response.status_code, 202)
+        self.assertTrue(updated_user.is_active)
+        self.assertEqual(changes.count(), 1)
+
+    def test_deactivate_user(self):
```
Comment (snopoke) [inline corehq/apps/api/tests/user_resources.py:360]: please split this test into 2, one for the normal user and one for the location user.

File: corehq/apps/api/resources/v0_5.py [hunk 3]
```diff
@@ -326,6 +331,40 @@ def _update(cls, bundle, user_change_logger=None):
...
+    @location_safe
+    def activate_user(self, request, **kwargs):
+        return self._modify_user_status(request, **kwargs, active=True)
+
+    @location_safe
+    def deactivate_user(self, request, **kwargs):
+        return self._modify_user_status(request, **kwargs, active=False)
+
+    def _modify_user_status(self, request, active, **kwargs):
+        self.is_authenticated(request)
```
Comment (snopoke) [inline corehq/apps/api/resources/v0_5.py:351]: This should also do a method and throttle check:
```suggestion
        self.method_check(request, allowed=['post'])
        self.is_authenticated(request)
        self.throttle_check(request)
```

File: corehq/apps/api/resources/v0_5.py [hunk 4]
```diff
@@ -326,6 +331,40 @@ def _update(cls, bundle, user_change_logger=None):
...
+        except ModifyUserStatusException as e:
+            raise BadRequest(_(str(e)))
+
+        user.is_active = active
+        user.save(spawn_task=True)
+        log_user_change(by_domain=request.domain, for_domain=user.domain,
+                        couch_user=user, changed_by_user=request.couch_user,
+                        changed_via=USER_CHANGE_VIA_API, fields_changed={'is_active': user.is_active})
+
+        return self.create_response(request, {}, response_class=http.HttpAccepted)
```
Comment (snopoke) [inline corehq/apps/api/resources/v0_5.py:366]: Log throttle access should be called for successful requests:

```suggestion
        self.log_throttled_access(request)
        return self.create_response(request, {}, response_class=http.HttpAccepted)
```

**Review by snopoke — COMMENTED** [review]
Generally looks good but a few changes required.

File: corehq/apps/api/resources/v0_5.py [hunk 5]
```diff
@@ -137,7 +142,7 @@
     v0_4,
 )
 from .pagination import DoesNothingPaginator, NoCountingPaginator, response_for_cursor_based_pagination
-
+from ...users.exceptions import ModifyUserStatusException
```
[the remark this answers is not in the document]
Comment (szymonbalasz) [reply corehq/apps/api/resources/v0_5.py:145]: Thanks, didn't check the auto-import there. 100% agree

**Comment by szymonbalasz** [comment]
> Generally looks good but a few changes required.

Thanks for the help! Have made updates as requested

**Comment by szymonbalasz** [comment]
@snopoke Just a friendly reminder of this PR. I also have another PR https://github.com/dimagi/commcare-hq/pull/34634 - but can't assign a reviewer - can you also take a look?

**Review by snopoke — APPROVED** [review]

**Pull request merged**